### ISSUE TITLE
[sampling] top_p can be deleted from both sampler paths without failing a single test

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1300]
## Summary

Both `top_p` implementations can be deleted entirely and the whole test suite stays green: no test in the repo ever passes a `top_p` that truncates a non-negligible tail.

## Repro

Commit: `ad067d76`
Mutants: `M20` (multiblock sampler), `M21` (CUB sampler) — both replace the nucleus cutoff with a never-taken branch:

```diff
--- a/src/compute/sampling_topk_topp.cu
+++ b/src/compute/sampling_topk_topp.cu
     for (int i = 0; i < top_k; ++i) {
         cumsum += s_val[i];
-        if (cumsum >= top_p) {
+        if (false && cumsum >= top_p) {
```

Command: `tools/mutation/run.py --only M20,M21`
Environment: RTX 5090, CUDA 13.3.1, `build-dev` (Release), full local suite (all 8 gtest binaries, every `IMP_TEST_MODEL*` set).

## Expected vs Actual

Expected: `SamplingTest.TopPFiltering` fails — it is the designated nucleus-sampling test.
Actual: both mutants **SURVIVED** the full suite. 0/2 killed.

## Evidence

`tests/test_sampling.cu:164` — the test documents one thing and exercises another:

```cpp
// Two tokens with equal high logits, rest negligible.
// top_p=0.5 should restrict to just one of the top tokens
constexpr int V = 10;
std::vector<float> logits(V, -100.0f);
logits[2] = 5.0f;
logits[7] = 5.0f;
...
int32_t token = sample_topk_topp(d_logits, /*top_k=*/128, /*top_p=*/0.99f, ...);
EXPECT_TRUE(token == 2 || token == 7);
```

The comment says `top_p=0.5`; the call passes `0.99`. With the other eight logits at `-100.0` the tail mass is ~e^-105, so the assertion holds whether or not the nucleus filter runs.

Census of every `top_p` value in the C++ suite:

```
$ rg -o 'top_p=\*/[0-9.]+f' tests/test_sampling.cu | sort | uniq -c
      2 top_p=*/0.95f
      1 top_p=*/0.99f
      6 top_p=*/1.0f      # a no-op
```

Plus `top_p = 1.0f` in five e2e fixtures and `0.8` in a config-parsing test. The CUB-path test (`SamplerCubPath.EachCallSamplesFromItsOwnLogitsNotThePreviousCalls`, `tests/test_sampling.cu:416`) uses `top_p=0.95` with one token at logit `30.0` and 151 935 at `-20.0` — again ~all mass on the winner.

`SamplingTest.SamplingDistribution` (`:184`) is the only test with a spread distribution, and it passes `top_p=1.0`.

## Why the test suite missed it

Escape class: **E6** — the test exists, is pointed at the right function, and its inputs cannot express the fault.

`SamplingTest.TopPFiltering` is the test that should have caught this. It fails to because its logit vector makes nucleus truncation a no-op at any `top_p >= ~0.5`.

## Severity

S1 — silent wrong output. A user setting `top_p=0.3` for a focused completion would receive samples from the full top-k with no error and no observable signal.

## Suggested test

One test, both paths, no new machinery:

- Build a spread distribution, e.g. softmax over logits `{3.0, 2.5, 2.0, 1.5, 1.0, ...}` so the top-3 carry ~0.6 of the mass.
- Sample 200 seeds at `top_p=0.55` and assert the **set** of returned tokens is a subset of the analytically-computed nucleus, and that the excluded-but-high-probability token (rank 4) never appears.
- Parameterise over `top_k = 50` (multiblock path) and `top_k = 200` (CUB path, `> SAMPLE_MAX_TOP_K = 128`) so both implementations are covered.

The nucleus set is computable in the test from the same logits, so this is an oracle test, not a golden.

Found by the test-hardening mutation baseline; see `docs/audit/MUTATION_BASELINE.md` and `docs/audit/ESCAPE_ANALYSIS.md`.

**Comment by kekzl** [comment]
Closed by the tests in #1306.

`SamplingTest.TopPTruncatesMultiblockPath` / `…Shifted` and `…CubPath` / `…Shifted` give `top_p` a value that truncates a real tail, with the nucleus computed in-test as an oracle and a control assertion that the same seeds *do* reach outside it at `top_p=1.0`. Both mutants (M20 multiblock, M21 CUB) are killed; verified by `tools/mutation/recheck.sh`.

That control assertion is also what surfaced #1305 — the CUB arm was returning token 0 for every seed — which is now fixed in #1322.
